### PR TITLE
Spark ignition utility

### PR DESCRIPTION
**Pull request by ThomasHowarth**
This PR adds the capability to add a source term to the energy equation, termed "spark".  The energy is deposited in a discrete spherical region at a particular time, for a specified duration - some discussion could be had about making these continuous functions, rather than being clipped in this way. Multiple sparks can be defined, and inputs have been added to documentation. Addresses some discussions had [here](https://github.com/AMReX-Combustion/PeleLMeX/discussions/384).

### CODE REVIEW

File: Source/PeleLMeX_Forces.cpp [hunk 1]
```diff
@@ -171,3 +171,70 @@ PeleLM::getVelForces(
...
+                 rho_a, rhoY_a, temp_src_a, rhoh_src_a, eos,
+                 dx] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
+          Real dist_to_center = std::sqrt(AMREX_D_TERM(
+            (i - spark_idx[0]) * (i - spark_idx[0]) * dx[0] * dx[0],
+            +(j - spark_idx[1]) * (j - spark_idx[1]) * dx[1] * dx[1],
+            +(k - spark_idx[2]) * (k - spark_idx[2]) * dx[2] * dx[2]));
+          if (dist_to_center < spark_radius) {
+            // probably doesn't actually
+            // contribute to anything
+            temp_src_a(i, j, k) = spark_temp / spark_duration;
```
Comment (baperry2) [inline Source/PeleLMeX_Forces.cpp]: Yes, this is not needed so you can remove

File: Source/PeleLMeX_Forces.cpp [hunk 2]
```diff
@@ -171,3 +171,70 @@ PeleLM::getVelForces(
...
+    }
+    Box domainBox = geom[lev].Domain();
+    // just a check
+    if (!domainBox.contains(spark_idx)) {
+      Warning(m_spark[n] + " not in domain!");
+      continue;
+    }
+    auto* ldata_p = getLevelDataPtr(lev, a_timestamp);
+    auto eos = pele::physics::PhysicsType::eos();
+    for (MFIter mfi(*(m_extSource[lev]), TilingIfNotGPU()); mfi.isValid();
```
Comment (baperry2) [inline Source/PeleLMeX_Forces.cpp]: Add OMP pragma similar to other ParallelFors in this file:
```
#ifdef AMREX_USE_OMP
#pragma omp parallel if (Gpu::notInLaunchRegion())
#endif
```

Or, the better option would be to convert this into the newer version of ParallelFor that operates directly on a MultiFab without MFIter (see, e.g, https://github.com/AMReX-Combustion/PeleLMeX/blob/8dccce90dfba3a1000a1c1602e701a06a0d09b7b/Source/PeleLMeX_Advection.cpp#L749, and note the `Gpu::StreamSynchronize` at the end of the loop)

File: Source/PeleLMeX_Advance.cpp [hunk 3]
```diff
@@ -74,6 +74,9 @@ PeleLM::Advance(int is_initIter)
 
   for (int lev = 0; lev <= finest_level; lev++) {
     m_extSource[lev]->setVal(0.);
+    if (m_n_sparks > 0) {
+      addSpark(lev, AmrNewTime);
```
Comment (baperry2) [inline Source/PeleLMeX_Advance.cpp]: Is this needed? We don't have an equivalent source term call for soot/spray/rad.

**Review by baperry2 — CHANGES_REQUESTED** [review]
Mostly looks good other than the few comments I added.

File: Source/PeleLMeX_Forces.cpp (hunk 2 above)
Comment (ThomasHowarth) [reply Source/PeleLMeX_Forces.cpp]: I've made this change - I assume there's a lot of functions lying around that need this change...

File: Source/PeleLMeX_Advance.cpp (hunk 3 above)
Comment (ThomasHowarth) [reply Source/PeleLMeX_Advance.cpp]: Doesn't look like it - removed.

File: Source/PeleLMeX_Forces.cpp (hunk 2 above)
Comment (baperry2) [reply Source/PeleLMeX_Forces.cpp]: Yeah there probably are. The multifab level ParFor can be a bit more efficient on GPU and simplifies the code (not needing MFiter and OMP pragmas), but it isn't a big enough deal to be a priority to go back and switch the old ones. Lucas switched a bunch of them over quite a while ago in potentially higher impact places.

**Review by baperry2 — APPROVED** [review]

**Pull request merged**